mono/mini/mini-posix.c

```diff
@@ -521,120 +504,58 @@ mono_runtime_cleanup_handlers (void)
 	free_saved_signal_handlers ();
 }
 
-#ifdef HAVE_LINUX_RTC_H
-#include <linux/rtc.h>
-#include <sys/ioctl.h>
-#include <fcntl.h>
-static int rtc_fd = -1;
+static MonoNativeThreadId sampling_thread;
+static volatile gint32 sampling_thread_running;
 
-static int
-enable_rtc_timer (gboolean enable)
+static mono_native_thread_return_t
+sampling_thread_func (void *data)
 {
-	int flags;
-	flags = fcntl (rtc_fd, F_GETFL);
-	if (flags < 0) {
-		perror ("getflags");
-		return 0;
-	}
-	if (enable)
-		flags |= FASYNC;
-	else
-		flags &= ~FASYNC;
-	if (fcntl (rtc_fd, F_SETFL, flags) == -1) {
-		perror ("setflags");
-		return 0;
+	/*
+	 * This gives us access to lock-free data structures that use hazard
+	 * pointers, e.g. the LLS (used for the thread list).
+	 */
+	mono_thread_info_register_small_id ();
+
+	MonoNativeThreadId me = mono_native_thread_id_get ();
+	guint64 us = 1000000 / mono_profiler_get_sampling_rate () - 1;
+
+	while (InterlockedRead (&sampling_thread_running)) {
+		FOREACH_THREAD_SAFE (info) {
+			if (mono_thread_info_get_tid (info) == me)
+				continue;
+
+			mono_threads_pthread_kill (info, SIGPROF);
+		} FOREACH_THREAD_SAFE_END
+
+		mono_thread_info_usleep (us);
```

Comment:
OK, I'm certainly not opposed to introducing new functions for this use case. I'll do that instead.

I feel like we ought to still discuss the confusing situation surrounding the sleep functions in mono-threads though. What (mis?)led me to believe that absolute time based waiting in `mono_thread_info_usleep` would be fine is the fact that `mono_thread_info_sleep` does exactly that. When it can, anyway; on some platforms, it falls back to `nanosleep` instead of `clock_nanosleep`, and if it's an interruptible sleep, it may use `sem_timedwait` which, as I understand it, is based on real time rather than a monotonic clock. Then there's the fact that the way it calls `nanosleep` seems to misuse the `rem` argument.

I guess it's just not very clear what a user of these two functions should be expecting from them.